File: java/src/test/java/mastermind/MastermindTest.java
package mastermind;

import static org.assertj.core.api.Assertions.*;

import java.util.List;

import org.junit.jupiter.api.Test;

class MastermindTest {
  
  @Test
  void shouldFailWithUnsupportedColor() {
    Mastermind mastermind = new Mastermind(fullRed());

    assertThat(mastermind.guess(List.of("Blue", "Blue", "Blue", "Brown"))).isEqualTo("One or more colors are unsupported");
  }

  @Test
  void shouldNotBeWellPlacedForRedSecretAndBlueGuess() {
    Mastermind mastermind = new Mastermind(fullRed());

    assertThat(mastermind.guess(List.of("Blue", "Blue", "Blue", "Blue"))).isEqualTo("[0,0]");
  }

  private List<String> fullRed() {
    return List.of("Red", "Red", "Red", "Red");
  }

  @Test
  void shouldSucceed() {
    Mastermind mastermind = new Mastermind(fullRed());
    var fullBlue = List.of("Blue", "Blue", "Blue", "Blue");
    mastermind.guess(fullBlue);
    mastermind.guess(fullBlue);
    mastermind.guess(fullBlue);
    mastermind.guess(fullBlue);
    mastermind.guess(fullBlue);
    mastermind.guess(fullBlue);
    mastermind.guess(fullBlue);
    mastermind.guess(fullBlue);
    mastermind.guess(fullBlue);

    assertThat(mastermind.guess(fullRed())).isEqualTo("Victory");
  }
  
  @Test
  void shouldBeSuccessfullAfterVictory() {
    Mastermind mastermind = new Mastermind(fullRed());
    var fullBlue = List.of("Blue", "Blue", "Blue", "Blue");
    mastermind.guess(fullBlue);
    mastermind.guess(fullRed());
    
    assertThat(mastermind.guess(fullRed())).isEqualTo("Victory");
  }
  
  @Test
  void shouldLooseAfterTenFailedAttempts() {
    Mastermind mastermind = new Mastermind(fullRed());
    
    var fullBlue = List.of("Blue", "Blue", "Blue", "Blue");
    mastermind.guess(fullBlue);
    mastermind.guess(fullBlue);
    mastermind.guess(fullBlue);
    mastermind.guess(fullBlue);
    mastermind.guess(fullBlue);
    mastermind.guess(fullBlue);
    mastermind.guess(fullBlue);
    mastermind.guess(fullBlue);
    mastermind.guess(fullBlue);
    
    assertThat(mastermind.guess(fullBlue)).isEqualTo("Failed");
  }
  
  @Test
  void shouldLooseAfterTenFailedAttemptsEvenIfNextAreSuccessfull() {
    Mastermind mastermind = new Mastermind(fullRed());
    
    var fullBlue = List.of("Blue", "Blue", "Blue", "Blue");
    mastermind.guess(fullBlue);
    mastermind.guess(fullBlue);
    mastermind.guess(fullBlue);
    mastermind.guess(fullBlue);
    mastermind.guess(fullBlue);
    mastermind.guess(fullBlue);
    mastermind.guess(fullBlue);
    mastermind.guess(fullBlue);
    mastermind.guess(fullBlue);
    mastermind.guess(fullBlue);
    
    assertThat(mastermind.guess(fullRed())).isEqualTo("Failed");
    assertThat(mastermind.guess(fullRed())).isEqualTo("Failed");
  }

  @Test
  void shouldNotBeWellPlacedForRedSecretAndRedBlueGuess() {
    Mastermind mastermind = new Mastermind(fullRed());

    assertThat(mastermind.guess(List.of("Red", "Blue", "Blue", "Blue"))).isEqualTo("[1,0]");
  }

  @Test
  void shouldNotBeWellPlacedForRedSecretAndRedBlueGuess2() {
    Mastermind mastermind = new Mastermind(List.of("Red", "Green", "Green", "Green"));

    assertThat(mastermind.guess(List.of("Blue", "Blue", "Red", "Blue"))).isEqualTo("[0,1]");
  }

  @Test
  void shouldNotBeWellPlacedForRedSecretAndRedBlueGuess4() {
    Mastermind mastermind = new Mastermind(List.of("Green", "Green", "Green", "Green"));

    assertThat(mastermind.guess(List.of("Green", "Green", "Red", "Blue"))).isEqualTo("[2,0]");
  }

  @Test
  void shouldNotBeWellPlacedForRedSecretAndRedBlueGuess5() {
    Mastermind mastermind = new Mastermind(List.of("Green", "Green", "Red", "Red"));

    assertThat(mastermind.guess(List.of("Red", "Red", "Blue", "Blue"))).isEqualTo("[0,2]");
  }

  @Test
  void shouldNotBeWellPlacedForRedSecretAndRedBlueGuess6() {
    List<String> secret = List.of("Green", "Blue", "Yellow", "Red");
    Mastermind mastermind = new Mastermind(secret);
    List<String> reversed = secret.reversed();

    assertThat(mastermind.guess(reversed)).isEqualTo("[0,4]");
  }
}

File: php/tests/Mastermind/MastermindTest.php
<?php

use App\Mastermind\Mastermind;

function fullRed()
{
    return ["Red", "Red", "Red", "Red"];
}

describe("Mastermind game", function()  {
    it("should fail with unsupported color", function() {
        $mastermind = new Mastermind(fullRed());

        expect($mastermind->guess(["Blue", "Blue", "Blue", "Brown"]))->toBe(
            "One or more colors are unsupported",
        );
    });
	it("should found nothing", function() {
        $mastermind = new Mastermind(fullRed());

        expect($mastermind->guess(["Blue", "Blue", "Blue", "Blue"]))->toBe("[0,0]");
    });
	it("should have one well placed color", function() {
        $mastermind = new Mastermind(fullRed());

        expect($mastermind->guess(["Red", "Blue", "Blue", "Blue"]))->toBe("[1,0]");
    });
	it("should have one existing color", function() {
        $mastermind = new Mastermind(["Red", "Green", "Green", "Green"]);

        expect($mastermind->guess(["Blue", "Blue", "Red", "Blue"]))->toBe("[0,1]");
    });
	it("should have two well placed colors", function() {
        $mastermind = new Mastermind(["Green", "Green", "Green", "Green"]);

        expect($mastermind->guess(["Green", "Green", "Red", "Blue"]))->toBe(
            "[2,0]",
        );
    });
	it("should found two colors", function() {
        $mastermind = new Mastermind(["Green", "Green", "Red", "Red"]);

        expect($mastermind->guess(["Red", "Red", "Blue", "Blue"]))->toBe("[0,2]");
    });
	it("should found four colors", function() {
        $secret = ["Green", "Blue", "Yellow", "Red"];
        $mastermind = new Mastermind($secret);
        $reversed = array_reverse($secret);

        expect($mastermind->guess($reversed))->toBe("[0,4]");
    });
	it("should win in ten rounds", function() {
        $mastermind = new Mastermind(fullRed());
        $fullBlue = ["Blue", "Blue", "Blue", "Blue"];

        $mastermind->guess($$fullBlue);
        $mastermind->guess($$fullBlue);
        $mastermind->guess($$fullBlue);
        $mastermind->guess($$fullBlue);
        $mastermind->guess($$fullBlue);
        $mastermind->guess($$fullBlue);
        $mastermind->guess($$fullBlue);
        $mastermind->guess($$fullBlue);
        $mastermind->guess($$fullBlue);

        expect($mastermind->guess(fullRed()))->toBe("Victory");
    });
	it("should win in two rounds", function() {
        $mastermind = new Mastermind(fullRed());
        $fullBlue = ["Blue", "Blue", "Blue", "Blue"];

        $mastermind->guess($fullBlue);
        $mastermind->guess(fullRed());

        expect($mastermind->guess(fullRed()))->toBe("Victory");
    });
	it("should failed at round ten", function() {
        $mastermind = new Mastermind(fullRed());
        $fullBlue = ["Blue", "Blue", "Blue", "Blue"];
        $mastermind->guess($fullBlue);
        $mastermind->guess($fullBlue);
        $mastermind->guess($fullBlue);
        $mastermind->guess($fullBlue);
        $mastermind->guess($fullBlue);
        $mastermind->guess($fullBlue);
        $mastermind->guess($fullBlue);
        $mastermind->guess($fullBlue);
        $mastermind->guess($fullBlue);

        expect($mastermind->guess($fullBlue))->toBe("Failed");
    });
	it("should failed even if an other guess is correct", function() {
        $mastermind = new Mastermind(fullRed());
        $fullBlue = ["Blue", "Blue", "Blue", "Blue"];
        $mastermind->guess($fullBlue);
        $mastermind->guess($fullBlue);
        $mastermind->guess($fullBlue);
        $mastermind->guess($fullBlue);
        $mastermind->guess($fullBlue);
        $mastermind->guess($fullBlue);
        $mastermind->guess($fullBlue);
        $mastermind->guess($fullBlue);
        $mastermind->guess($fullBlue);
        $mastermind->guess($fullBlue);

        expect($mastermind->guess(fullRed()))->toBe("Failed");
        expect($mastermind->guess(fullRed()))->toBe("Failed");
    });
});

